### PR TITLE
Create undeclared-action-input.ql

### PR DESCRIPTION
**Pull request by robertbrignull**
Adds a new CodeQL query for detecting when we are referencing action input parameters (by `core.getInput(...)`) from actions where those inputs are not defined. This can easily happen when code is reused between actions, and affects us more than most people because we have multiple actions in the same repo.

The alerts are visible at https://github.com/github/codeql-action/security/code-scanning?query=ref%3Arefs%2Fheads%2Fundeclared-action-input

All the alerts appear to be true positives to me and it's found all the alerts I expected it to find.

(Here's a screenshot for posterity)
results2

### CODE REVIEW

File: queries/undeclared-action-input.ql [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    result.getFile().getRelativePath() = getRootNode().
+      (YAMLMapping).lookup("runs").
+      (YAMLMapping).lookup("main").
+      (YAMLString).getValue().regexpReplaceAll("\\.\\./lib/(.*)\\.js", "src/$1.ts") and
+    result.getName() = "run"
+  }
+}
+
+Expr getAFunctionChildExpr(Function f) {
+  result = f.getBody().getAChildStmt*().getAChildExpr*()
```
Comment (robertbrignull) [inline queries/undeclared-action-input.ql:39]: I'm not very knowlegeable about the javascript QL libraries so might need some help from the team. Perhaps @asgerf if you have a chance. Do the `getAFunctionChildExpr` and `calledBy` predicates look correct to you, or are they even necessary and maybe there's something in the library that can do this more easily?

**Comment by joshhale** [comment]
CodeQL is failing because of the alerts detected by this PR, but I think they can/should be addressed separately.

File: queries/undeclared-action-input.ql (hunk 1 above)
Comment (asgerf) [reply queries/undeclared-action-input.ql:39]: `getContainer` is probably what you're looking for:
```suggestion
  result.getContainer() = f
```

There's a subtle difference there in how nested functions are handled, however (yours would include the body of arrow functions).

To handle nested functions, I'd also add this to `calledBy`:
```codeql
Function calledBy(Function f) {
  ...
  or
  result.getEnclosingContainer() = f // assume outer function causes inner function to be called
}
```

Comment (robertbrignull) [reply queries/undeclared-action-input.ql]: Thanks. Didn't know about that predicate but that makes sense.

**Comment by robertbrignull** [comment]
Queries still seem to work. It finds all the alerts that I expected it to.

The "Code scanning / CodeQL" check is marked as red because this is "introducing" alerts, but this is fine and we've decided we want to fix them in separate PRs.

**Pull request merged**